### PR TITLE
[rom_cntrl, dv] Assertion to validate rom controller outputs to KMAC

### PR DESCRIPTION
**Pull request by prajwalaputtappa**
Following assertions have been added in this commit:

1. Assertion to ensure kmac_data_o.valid is never aserted after
pwrmgr_data_o.done is asserted.

Signed-off-by: Prajwala Puttappa <prajwalaputtappa@lowrisc.org>

### CODE REVIEW

File: hw/ip/rom_ctrl/dv/sva/kmac_valid_done_assertion.sv
Comment (weicaiyang) [inline hw/ip/rom_ctrl/dv/sva/kmac_valid_done_assertion.sv:5]: not sure if we should have one module for one assertion or just create one module for all the rom_ctrl assertions except tlul.

**Review by weicaiyang — COMMENTED** [review]
overall LGTM with a question

File: hw/ip/rom_ctrl/dv/sva/kmac_valid_done_assertion.sv
Comment (rswarbrick) [reply hw/ip/rom_ctrl/dv/sva/kmac_valid_done_assertion.sv:5]: I think I'd suggest one module for each set of interface assertions. BUT: This is an assertion about behaviour between two interfaces. @prajwalaputtappa: Could you change things to just plonk the assertion into `tb.sv`? I think that might be a bit clearer.

Comment (prajwalaputtappa) [reply hw/ip/rom_ctrl/dv/sva/kmac_valid_done_assertion.sv]: I've moved the assertion to the end of tb.sv

File: hw/ip/rom_ctrl/dv/tb.sv
Comment (rswarbrick) [inline hw/ip/rom_ctrl/dv/tb.sv:96]: What we want to assert is that once we've finished an operation (and asserted `pwrmgr_data.done`), we'll never assert `kmac_data_out.valid` in future.

This code does imply that, but only because we also know that `pwrmgr_data.done` is never de-asserted. I think the comment needs expanding to explain what we're trying to assert ("We don't send any requests to KMAC once we've asserted `done` on the power manager interface") and also why this code checks the high-level requirement ("we've got an assertion bound into the interface to check that the done signal never drops").

HOWEVER: We could actually assert a stronger statement, that we never send out any KMAC requests once we've had a digest. Maybe that would be better? We could also assert that we don't signal that we're done until we've had the digest from KMAC.

File: hw/ip/rom_ctrl/dv/tb.sv
Comment (rswarbrick) [inline hw/ip/rom_ctrl/dv/tb.sv:108]: I think that we can express something rather stronger without much work. We're really saying that the following events happen in a fixed order:

A. We send some requests to KMAC (`kmac_data_out.valid`)
B. We get a response from KMAC (`kmac_data_in.done`)
C. We assert done to pwrmgr (`pwrmgr_data.done`)

Event A happens several times. Events B and C should happen once each. So I think we can model things with the following assertions:
1. Once B has happened, it doesn't happen again.
2. Once C has happened, it doesn't happen again.
3. Once B has happened, A doesn't happen.
4. Once C has happened, B doesn't happen. *This one's not actually needed if you have (5) + (1)*
5. B happens before C.

Unfortunately, (5) needs an extra signal (you could do it with something like `!C until B` in an initial assertion, but that's not possible with our `ASSERT_INIT` macro, which expects an immediate expression rather than an arbitrary property).

I think the result is that we end up with:
```suggestion
  // Only get one KMAC response
  `ASSERT(JustOneKmacResponse_A,
          kmac_data_in.done |=> always !kmac_data_in.done,
          clk, rst_n)

  // Once we signal done to pwrmgr, the done signal stays high and the good signal stays stable.
  `ASSERT(PwrmgrDoneOneWay_A,
          $rose(pwrmgr_data.done == prim_mubi_pkg::MuBi4True) |=>
          always $stable(pwrmgr_data),
          clk, rst_n)

  // Don't send any KMAC requests once we've had the response
  `ASSERT(KmacNotOutAfterIn_A,
          kmac_data_in.done |=> always !kmac_data_out.valid,
          clk, rst_n)

  // We see a response from KMAC before we assert that we're done to pwrmgr
  logic seen_kmac_response;
  always_ff @(posedge clk or negedge rst_n) begin
    if (!rst_n) begin
      seen_kmac_response <= 1'b0;
    end else begin
      seen_kmac_response <= seen_kmac_response | kmac_data_in.done;
    end
  end
  `ASSERT(KmacResponseBeforePwmgrDone_A,
          (pwrmgr_data.done == prim_mubi_pkg::MuBi4True) |-> seen_kmac_response,
          clk, rst_n)
```

An alternative way of writing that last one that doesn't need the auxiliary signal would be something like:
```systemverilog
  initial begin
    KmacResponseBeforePwmgrDone_A:
      assert property (@(posedge clk) disable iff (!rst_n)
                       (pwrmgr_data.done == prim_mubi_pkg::MuBi4False) until kmac_data_in.done);
  end
```
@sriyerg, @cindychip: Do you have any cunning tricks for writing this with our existing macro set?

Comment (weicaiyang) [reply hw/ip/rom_ctrl/dv/tb.sv:108]:
```
  initial begin
    KmacResponseBeforePwmgrDone_A:
      assert property (@(posedge clk) disable iff (!rst_n)
                       (pwrmgr_data.done == prim_mubi_pkg::MuBi4False) until kmac_data_in.done);
  end
```

For above case, can we use this?
```
`ASSERT(KmacResponseBeforePwmgrDone_A,
                pwrmgr_data.done == prim_mubi_pkg::MuBi4False || kmac_data_in.done)
```

Comment (rswarbrick) [reply hw/ip/rom_ctrl/dv/tb.sv:108]: That assertion will be false as soon as `pwrmgr_data.done` goes true (because `kmac_data_in.done` is a single-cycle pulse).

Comment (weicaiyang) [reply hw/ip/rom_ctrl/dv/tb.sv:108]: Seems like you don't need to check it after kmac_data_in.done
We could add a helper logic.

```
logic kmac_done_occured.

always_comb begin
  if (!rst_n) kmac_done_asserted = 0;
  else if (kmac_data_in.done) kmac_done_occured = 1;
end

`ASSERT(KmacResponseBeforePwmgrDone_A,
                pwrmgr_data.done == prim_mubi_pkg::MuBi4False || kmac_done_occured)
```
there may be some other better approach too.

Comment (prajwalaputtappa) [reply hw/ip/rom_ctrl/dv/tb.sv]: Thanks @rswarbrick and @weicaiyang. I've made the necessary changes.

File: hw/ip/rom_ctrl/dv/tb.sv
Comment (rswarbrick) [inline hw/ip/rom_ctrl/dv/tb.sv:110]: Sorry, I realise that this was my suggestion but this would probably be easier to read as `pwrmgr_data.done != prim_mubi_pkg::MuBi4False |-> kmac_done_occurred`. (I think I was doing the `A |-> B iff B || !A` thing a bit too enthusiastically).

**Review by rswarbrick — COMMENTED** [review]
Looks good to me, but probably needs approval from a DV engineer who didn't write the design.

File: hw/ip/rom_ctrl/dv/tb.sv
Comment (prajwalaputtappa) [reply hw/ip/rom_ctrl/dv/tb.sv]: Changes made!
@weicaiyang, can yo please review and approve if all is well?

**Review by weicaiyang — APPROVED** [review]
LGTM, just learned we can use `always` in SVA

**Comment by rswarbrick** [comment]
> LGTM, just learned we can use always in SVA

Yeah, it's neat. There's an embedding of LTL in SystemVerilog assertions, so all the G's, F's and U's carry over as you'd expect.

**Pull request merged**
